[Merged by Bors] - feat(linear_algebra/matrix/charpoly/eigs): det and trace are product and sum of eigenvalues

MohanadAhmed opened the pull request:
This adds two lemmas:
1. `det_eq_prod_roots_charpoly`: the determinant of a square matrix is the product of the characteristic polynomial roots of that matrix.
2. `trace_eq_sum_roots_charpoly`: the trace is similarly the sum of the characteristic polynomial roots.

These two lemmas are more commonly stated as trace is the sum of eigenvalues and determinant is the product of eigenvalues. Mathlib has already defined eigenvalues in [linear_algebra.eigenspace](https://leanprover-community.github.io/mathlib_docs/linear_algebra/eigenspace.html#module.End.eigenvalues) as the roots of the minimal polynomial of a linear endomorphism. These do not have correct multiplicity and cannot be used in the theorems above. Hence we express these theorems in terms of the roots of the characteristic polynomial directly.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

MohanadAhmed committed:
Det and Trace are product and sum of eigenvalues
Only algebraically closed fields

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,77 @@
+open linear_map module.End
+open_locale matrix big_operators
+
+lemma det_eq_prod_roots_charpoly (A : matrix n n R):
```
This name is backwards,
```suggestion
lemma prod_roots_charpoly_eq_det (A : matrix n n R) :
```

MohanadAhmed replied:
I think people will search for it the other way around using the determinant keyword. Perhaps we can switch the equation itself to say:
`A.det = (matrix.charpoly A).roots.prod`
Similarly the other one says:
`A.trace = (matrix.charpoly A).roots.sum`

eric-wieser replied:
Usually the convention is to put the simpler expression on the RHS, which is `det` or `trace`. When constructing a query searching mathlib it's expected that you apply this heuristic too.

MohanadAhmed replied:
You mean simpler in length? Or conceptually simpler?

How about these theorems in mathlib:
- [matrix.trace_eq_neg_charpoly_coeff](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix/charpoly/coeff.html#matrix.trace_eq_neg_charpoly_coeff)
- [matrix.det_eq_sign_charpoly_coeff](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix/charpoly/coeff.html#matrix.det_eq_sign_charpoly_coeff)
- [matrix.is_hermitian.det_eq_prod_eigenvalues](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix/spectrum.html#matrix.is_hermitian.det_eq_prod_eigenvalues)
- [matrix.is_hermitian.eigenvalues_eq](https://leanprover-community.github.io/mathlib_docs/linear_algebra/matrix/spectrum.html#matrix.is_hermitian.eigenvalues_eq)

I think the simpler side is the LHS in the lemmas mentioned above.

In the derivations I am imagining the user will probably consider the eigenvalues (charpoly roots) more simple as they allow him to do further manipulations. In that sense "determinant equals product of eigenvalues" is more like a "definition" you "unfold" to do more work with, than a simp theorem that gives a simplification on its right hand side. For example consider proving `det(A) < det(1 + A)` for a positive semidefinite matrix because every eigenvalue is increased by 1 on the RHS.

Anyways, I will defer to your experience on this one and reverse the lemma names.

eric-wieser replied:
Ah, those first two lemmas are a good counterargument. Feel free to use that order after all.

Just make sure the lemma name and statement match!

MohanadAhmed committed:
Update eigs.lean

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,77 @@
+- `det_eq_prod_roots_charpoly` determinant is the product of the roots of the characteristic
+polynomial.
+- `trace_eq_sum_roots_charpoly` trace of a matrix is the sum of the roots of the characteristic
+polynomial.
```
```suggestion

In algebraically closed fields, we show:
* `matrix.det_eq_prod_roots_charpoly`: the determinant is the product of the roots of the characteristic
polynomial.
* `matrix.trace_eq_sum_roots_charpoly`: the trace is the sum of the roots of the characteristic
polynomial.

Note that over other fields such as ` ℝ`, these results can be used by using `A.map (algebra_map ℝ ℂ)` as the matrix,
and then applying `ring_hom.map_det`.
```

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,79 @@
+lemma prod_roots_charpoly_eq_det (A : matrix n n R) :
+  A.det = (matrix.charpoly A).roots.prod :=
+begin
+  by_cases hn: nonempty n,
```
```suggestion
  casesI is_empty_or_nonempty n,
```

eric-wieser replied:
Though you don't even need this case split at all with my proof below!

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,79 @@
+      rw [neg_one_eq_one_iff.2 h2, one_pow, one_mul],},
+
+    apply matrix.charpoly_monic,
+    exact is_alg_closed.splits A.charpoly, },
```
Case-splitting on char 2 is usually an indication that you're using the wrong lemmas!
```suggestion
    rw [det_eq_sign_charpoly_coeff, ← hdeg,
      polynomial.prod_roots_eq_coeff_zero_of_monic_of_split A.charpoly_monic
        (is_alg_closed.splits _),
      ← mul_assoc, ← pow_two, pow_right_comm, neg_one_sq, one_pow, one_mul] },
```

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,79 @@
+open_locale matrix big_operators
+
+lemma prod_roots_charpoly_eq_det (A : matrix n n R) :
+  A.det = (matrix.charpoly A).roots.prod :=
```
```suggestion
  (matrix.charpoly A).roots.prod = A.det :=
```
as discussed previously

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,79 @@
+lemma trace_eq_sum_roots_charpoly (A: matrix n n R) :
+  A.trace = (matrix.charpoly A).roots.sum :=
+begin
+  by_cases hn: (nonempty n),
```
```suggestion
  casesI is_empty_or_nonempty n,
```
and then swap around the cases below

MohanadAhmed replied:
OK I see so `casesI isempty_or_nonemtpy n` adds the lemma directly to whatever system that picks up the parameters in square brackets.

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Name reversal and other comments by eric on the PR

MohanadAhmed committed:
Line wrapping comments to 100 chars per line

MohanadAhmed committed:
Reverse lemma names again

MohanadAhmed committed:
A few more golfs

MohanadAhmed committed:
det lemma name order matches sttatement order

MohanadAhmed committed:
More golf

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,69 @@
+    rw [det_eq_sign_charpoly_coeff, ← hdeg,
+      polynomial.prod_roots_eq_coeff_zero_of_monic_of_split A.charpoly_monic
+        (is_alg_closed.splits _),
+      ← mul_assoc, ← pow_two, pow_right_comm, neg_one_sq, one_pow, one_mul], },
```
I'm pretty sure that this proof doesn't actually need `nonempty n`; so you can remove the case-split above.

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,69 @@
+    rw matrix.charpoly,
+    rw det_eq_one_of_card_eq_zero (fintype.card_eq_zero_iff.2 h),
+    rw polynomial.roots_one,
+    simp only [multiset.empty_eq_zero, multiset.sum_zero],},
```
Can you collapse these `rw`s into a single `rw`?

MohanadAhmed replied:
Is there a way to make this part into a term mode thing:
```
have fne := @fintype.card_ne_zero _ _ h,
rw ne.def at fne,
split_ifs,
refl,
```

I mean what is the term mode equivalent of split_ifs?

MohanadAhmed committed:
Erics suggested golfs

eric-wieser committed:
style fixes

eric-wieser committed:
cleanup

eric-wieser reviewed (approved):
maintainer merge

@Vierkantor, does this look ok to you? I'm not entirely sure whether `det_eq_sign_charpoly_coeff` and `trace_eq_neg_charpoly_coeff` claimed to be these result in the first place, and whether the results in this file should live next to those ones.

Vierkantor commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,59 @@
+
+import data.polynomial.basic
+import field_theory.is_alg_closed.basic
+import algebra.char_p.two
```
The import of `algebra.char_p.two` caught my attention: why do we need it?

eric-wieser replied:
This is a vestigial import from a much less direct proof that I golfed; good catch that I forgot to remove it!

eric-wieser replied:
```suggestion
```

Vierkantor commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,59 @@
+* `matrix.trace_eq_sum_roots_charpoly`: the trace is the sum of the roots of the characteristic
+  polynomial.
+
+Note that over other fields such as ` ℝ`, these results can be used by using
```
```suggestion
Note that over other fields such as `ℝ`, these results can be used by using
```

Vierkantor commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,59 @@
+polynomial directly.
+-/
+variables {n : Type*} [fintype n] [decidable_eq n]
+variables {R : Type*} [field R] [is_alg_closed R]
```
We only really need that one specific polynomial splits, so let's weaken this hypothesis.
```suggestion
variables {R : Type*} [field R] (h : (matrix.charpoly A).splits)
```

MohanadAhmed replied:
This seems to require and extra argument `(i : K →+* L)`. It says this "bundled ring homomorphisms".[ring_hom](https://leanprover-community.github.io/mathlib_docs/algebra/hom/ring.html#ring_hom).
I have no idea what that is or how to get one. As you can guess my lack of knowledge about this thing, makes me think this will make both lemmas much more difficult to use. !!!

Vierkantor reviewed (commented):
I think these results are worth including, since they are stated in a more "mathematical" form than `det_eq_sign_charpoly_coeff` and `trace_eq_neg_charpoly_coeff`. However, the new results involve somewhat more machinery, namely `is_alg_closed`, so the old versions should stay.

What I'd like to see is more integration with [`norm_gen_eq_prod_roots`](https://github.com/leanprover-community/mathlib/blob/6480bedf0354aaba0016eeefd726f7e3e5fc50aa/src/ring_theory/norm.lean#L123) and [`trace_gen_eq_sum_roots`](https://github.com/leanprover-community/mathlib/blob/6480bedf0354aaba0016eeefd726f7e3e5fc50aa/src/ring_theory/trace.lean#L231), since the proofs go through the same steps: applying `det_eq_sign_charpoly_coeff`/`trace_eq_neg_charpoly_coeff` and then relating those to the roots, assuming the polynomial is split. Do you think it's feasible to cut out that part of the proof into a new pair of lemmas?

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

MohanadAhmed committed:
Vierkantor and Erics Review

MohanadAhmed committed:
Two versions. The shorter named ones pick up algebraic closure instances and do the splitting themselves.

The longer names ones ...... well split hairs and prefer if you don't call them!!!

MohanadAhmed committed:
Space knitpick

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,77 @@
+  det_eq_prod_roots_charpoly_of_splits (is_alg_closed.splits A.charpoly)
+
+lemma trace_eq_sum_roots_charpoly [is_alg_closed R] : A.trace = (matrix.charpoly A).roots.sum :=
+  trace_eq_sum_roots_charpoly_of_splits (is_alg_closed.splits A.charpoly)
```
```suggestion
variables (A)

lemma det_eq_prod_roots_charpoly [is_alg_closed R] : A.det = (matrix.charpoly A).roots.prod :=
det_eq_prod_roots_charpoly_of_splits (is_alg_closed.splits A.charpoly)

lemma trace_eq_sum_roots_charpoly [is_alg_closed R] : A.trace = (matrix.charpoly A).roots.sum :=
trace_eq_sum_roots_charpoly_of_splits (is_alg_closed.splits A.charpoly)
```
since these have no `hAps` argument to infer `A` from.

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,77 @@
+    ← mul_assoc, ← pow_two, pow_right_comm, neg_one_sq, one_pow, one_mul],
+end
+
+lemma trace_eq_sum_roots_charpoly_of_splits (hAps : (matrix.charpoly A).splits (ring_hom.id R)) :
```
I think you can abbreviate this to
```suggestion
lemma trace_eq_sum_roots_charpoly_of_splits (hAps : A.charpoly.splits (ring_hom.id R)) :
```

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Erics suggestions

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,78 @@
+namespace matrix
+
+variables
+
```
```suggestion
```

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,75 @@
+Mathlib has already defined eigenvalues in `linear_algebra.eigenspace` as the roots of the minimal
+polynomial of a linear endomorphism. These do not have correct multiplicity and cannot be used in
+the theorems above. Hence we express these theorems in terms of the roots of the characteristic
+polynomial directly.
```
```suggestion
polynomial directly.

## TODO

The proofs of `det_eq_prod_roots_charpoly_of_splits` and
`trace_eq_sum_roots_charpoly_of_splits` closely resemble
`norm_gen_eq_prod_roots` and `trace_gen_eq_sum_roots` respectively, but the
dependencies are not general enough to unify them. We should refactor
`polynomial.prod_roots_eq_coeff_zero_of_monic_of_split` and
`polynomial.sum_roots_eq_next_coeff_of_monic_of_split` to assume splitting over an arbitrary map.
```

Vierkantor reviewed (approved):
Thanks for the changes!

I tried to unify the commonalities with `norm_gen_eq_prod_roots` and `trace_gen_eq_sum_roots` myself, but had a few issues since `prod_roots_eq_coeff_zero_of_monic_of_split` is not quite general enough. So I propose that we merge this now with this TODO comment, and then come back to this later (after the mathlib4 port, probably). @eric-wieser, do you agree?

bors d+

MohanadAhmed committed:
Update src/linear_algebra/matrix/charpoly/eigs.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

eric-wieser commented on src/linear_algebra/matrix/charpoly/eigs.lean:
```diff
@@ -0,0 +1,84 @@
+`norm_gen_eq_prod_roots` and `trace_gen_eq_sum_roots` respectively, but the
+dependencies are not general enough to unify them. We should refactor
+`polynomial.prod_roots_eq_coeff_zero_of_monic_of_split` and
+`polynomial.sum_roots_eq_next_coeff_of_monic_of_split` to assume splitting over an arbitrary map.
```
Why does `polynomial.splits` take a morphism at all? Doesn't `polynomial.splits_map_iff` demonstrate that it's useless and we should just use `(p.map f).splits` instead?
